keras_nlp/models/xlm_roberta/xlm_roberta_preprocessing.py

```diff
@@ -0,0 +1,199 @@
+                left-to-right manner and fills up the buckets until we run
+                out of budget. It supports an arbitrary number of segments.
+
+Examples:
+```python
+preprocessor = keras_nlp.models.XLMRobertaPreprocessor(proto="model.spm")
+
+# Tokenize and pack a single sentence directly.
+preprocessor("The quick brown fox jumped.")
+
+# Tokenize and pack a multiple sentence directly.
+preprocessor(("The quick brown fox jumped.", "Call me Ishmael."))
+
+# Map a dataset to preprocess a single sentence.
+features = ["The quick brown fox jumped.", "I forgot my homework."]
+labels = [0, 1]
+ds = tf.data.Dataset.from_tensor_slices((features, labels))
+ds = ds.map(
+    lambda x, y: (preprocessor(x), y),
+    num_parallel_calls=tf.data.AUTOTUNE,
+)
+
+# Map a dataset to preprocess a multiple sentences.
+first_sentences = ["The quick brown fox jumped.", "Call me Ishmael."]
+second_sentences = ["The fox tripped.", "Oh look, a whale."]
+labels = [1, 1]
+ds = tf.data.Dataset.from_tensor_slices(((first_sentences, second_sentences), labels))
+ds = ds.map(
+    lambda x, y: (preprocessor(x), y),
+    num_parallel_calls=tf.data.AUTOTUNE,
+)
+```
+"""
+
+
+class XLMRobertaPreprocessor(keras.layers.Layer):
+    def __init__(
+        self,
+        proto,
+        sequence_length=512,
+        truncate="round_robin",
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        proto = _handle_pretrained_tokenizer_arguments(proto)
+        self.tokenizer = SentencePieceTokenizer(proto=proto)
+
+        # Check for necessary special tokens.
+        start_token_id = 0
+        pad_token_id = 1
+        end_token_id = 2
+
+        self.packer = RobertaMultiSegmentPacker(
+            start_value=start_token_id,
+            end_value=end_token_id,
+            pad_value=pad_token_id,
+            truncate=truncate,
+            sequence_length=sequence_length,
+        )
+
+        self.pad_token_id = pad_token_id
+
+    def vocabulary_size(self):
+        """Returns the vocabulary size of the tokenizer."""
+        return self.tokenizer.vocabulary_size()
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "proto": self.tokenizer.proto,
+                "sequence_length": self.packer.sequence_length,
+                "truncate": self.packer.truncate,
+            }
+        )
+        return config
+
+    def call(self, inputs):
+        if not isinstance(inputs, (list, tuple)):
+            inputs = [inputs]
+
+        def _tokenize(x):
+            tokenized = self.tokenizer(x)
+            dtype = tokenized.dtype
+
+            # In the official SPM proto file, `[unk]`'s ID is 0. Replace that
+            # with 2. This will be changed to 3 (`[unk]`'s ID is 3 in the
+            # official implementation) after adding by 1.
+            tokenized = tf.where(
+                tf.equal(tokenized, tf.constant(0, dtype=dtype)),
```

Comment:
will tf not automatically convert `0` correctly? that's unfortunate if so

Reply:
Ohh, does it? I've faced some issues in the past...so, thought I had to typecast them. Will test it without typecasting

Reply:
Cool, it worked without typecasting!